backend/src/main/java/fk/prof/backend/model/association/impl/ZookeeperBasedBackendAssociationStore.java

```diff
@@ -0,0 +1,336 @@
+  }
+
+  /**
+   * De associates a backend from process group. Writes to in-memory store and zookeeper
+   * NOTE: Ensure that the backend detail instance being modified has been removed from available backend queue prior to calling this method
+   * @param backendDetail
+   * @param processGroup
+   * @throws Exception
+   */
+  private void deAssociateBackendWithProcessGroup(BackendDetail backendDetail, Recorder.ProcessGroup processGroup)
+      throws Exception {
+    Set<Recorder.ProcessGroup> associatedProcessGroups = new HashSet<>(backendDetail.getAssociatedProcessGroups());
+    associatedProcessGroups.remove(processGroup);
+
+    String zNodePath = getZNodePathForBackend(backendDetail.getBackendIPAddress());
+    ZookeeperUtil.writeZNode(curatorClient, zNodePath, BackendDetail.serializeProcessGroups(associatedProcessGroups), false);
+
+    backendDetail.deAssociateProcessGroup(processGroup);
+    processGroupToBackendLookup.remove(processGroup);
+  }
+
+  private List<BackendDetail> loadDataFromZookeeper() throws Exception {
+    List<BackendDetail> backends = new ArrayList<>();
+    List<String> backendIPAddresses = curatorClient.getChildren().forPath(backendAssociationPath);
+    for(String backendIPAddress: backendIPAddresses) {
+      String zNodePath = getZNodePathForBackend(backendIPAddress);
+      backends.add(new BackendDetail(backendIPAddress, reportingFrequencyInSeconds, maxAllowedSkips, ZookeeperUtil.readZNode(curatorClient, zNodePath)));
+    }
+
+    return backends;
+  }
+
+  private String getZNodePathForBackend(String backendIPAddress) {
+    return backendAssociationPath + "/" + backendIPAddress;
+  }
+
+  public static Builder newBuilder() {
+    return new Builder();
+  }
+
+  public static class Builder {
+    private CuratorFramework curatorClient;
+    private String backendAssociationPath;
+    private Integer reportingFrequencyInSeconds;
+    private Integer maxAllowedSkips = 2;
+    private Comparator<BackendDetail> backendPriorityComparator = new ProcessGroupCountBasedBackendComparator();
+
+    public Builder setCuratorClient(CuratorFramework curatorClient) {
+      this.curatorClient = curatorClient;
+      return this;
+    }
+
+    public Builder setBackendAssociationPath(String backendAssociationPath) {
+      this.backendAssociationPath = backendAssociationPath;
+      return this;
+    }
+
+    public Builder setReportingFrequencyInSeconds(int reportingFrequencyInSeconds) {
+      this.reportingFrequencyInSeconds = reportingFrequencyInSeconds;
+      return this;
+    }
+
+    public Builder setMaxAllowedSkips(int maxAllowedSkips) {
+      this.maxAllowedSkips = maxAllowedSkips;
+      return this;
+    }
+
+    public Builder setBackedPriorityComparator(Comparator<BackendDetail> backedPriorityComparator) {
+      this.backendPriorityComparator = backedPriorityComparator;
+      return this;
+    }
+
+    public ZookeeperBasedBackendAssociationStore build(Vertx vertx) throws Exception {
```

Comment:
This was not driven by usability, rather using builder pattern here just made the code more readable when instantiating this class.